### PR TITLE
Allow specifying the compare-array for insertIfNotExists()

### PR DESCRIPTION
**Pull request by nickvergessen**

### CODE REVIEW

File: lib/private/allconfig.php [hunk 1]
```diff
@@ -189,11 +189,22 @@ public function setUserValue($userId, $appName, $key, $value, $preCondition = nu
...
+			$this->connection->insertIfNotExist('*PREFIX*preferences', [
+				'configvalue'	=> $value,
+				'userid'		=> $userId,
+				'appid'			=> $appName,
+				'configkey'		=> $key,
+			], [
+				'configvalue'	=> $value,
+				'userid'		=> $userId,
+				'appid'			=> $appName,
+			]);
```
Comment (LukasReschke) [inline lib/private/allconfig.php]: What about only having to specify the array keys instead of duplicating the values as well?

Comment (nickvergessen) [inline lib/private/allconfig.php]: I guess that makes sense, yes

**Comment by DeepDiver1975** [comment]
@nickvergessen I see what you are trying to accomplish - bääh - we should have continue to talk about this days back ... well - shit happens.

**Comment by DeepDiver1975** [comment]
What I was thinking about while having a look at the implementation of insertIfNoExists():
What about handling the unique constraint violation not is a specific way - like simply throwing a UniqueContraintViolationException and let the caller decide on what to do.

**Comment by DeepDiver1975** [comment]
61e4fd7 - contains a unit test to show this approach basically works
 1e7d3ee - DBALException is throws from insertIfNotExists instead of HintException

File: lib/private/files/cache/storage.php [hunk 2]
```diff
@@ -35,9 +35,17 @@ public function __construct($storage) {
...
-			$this->numericId = \OC_DB::insertid('*PREFIX*storages');
+			$connection = \OC_DB::getConnection();
+			if ($connection->insertIfNotExist('*PREFIX*storages', ['id' => $this->storageId])) {
+				$this->numericId = \OC_DB::insertid('*PREFIX*storages');
+			} else {
+				$result = \OC_DB::executeAudited($sql, array($this->storageId));
+				if ($row = $result->fetchRow()) {
+					$this->numericId = $row['numeric_id'];
+				} else {
+					throw new \Exception('Storage exists when inserting and does not exist on select... go away');
```
Comment (nickvergessen) [inline lib/private/files/cache/storage.php]: This is temp code and must not happen

**Comment by nickvergessen** [comment]
> What I was thinking about while having a look at the implementation of insertIfNoExists():
What about handling the unique constraint violation not is a specific way - like simply throwing a UniqueContraintViolationException and let the caller decide on what to do.

I think we can/should do this, to make it possible to differ between a normal DB exception and this case. Shall I go ahead and implement that @DeepDiver1975

**Comment by DeepDiver1975** [comment]
> Shall I go ahead and implement that

 89be55a <- already done

File: lib/private/allconfig.php [hunk 3]
```diff
@@ -189,11 +189,18 @@ public function setUserValue($userId, $appName, $key, $value, $preCondition = nu
 			return;
 		}
 
-		$data = array($value, $userId, $appName, $key);
+		$affectedRows = 0;
 		if (!$exists && $preCondition === null) {
-			$sql  = 'INSERT INTO `*PREFIX*preferences` (`configvalue`, `userid`, `appid`, `configkey`)'.
-					'VALUES (?, ?, ?, ?)';
+			$this->connection->insertIfNotExist('*PREFIX*preferences', [
```
Comment (DeepDiver1975) [inline lib/private/allconfig.php]: this whole code block cries for a second method: InsertOrUpdate - but this is OT

File: lib/private/db/adaptersqlite.php [hunk 4]
```diff
@@ -18,38 +18,35 @@ public function fixupStatement($statement) {
 		return $statement;
 	}
 
-	public function insertIfNotExist($table, $input) {
+	/**
+	 * @param string $table
+	 * @param array $input
+	 * @param null $compare
```
Comment (DeepDiver1975) [inline lib/private/db/adaptersqlite.php]: array | null :question:

**Comment by DeepDiver1975** [comment]
From my POV this is ready to go - :high_five: @nickvergessen

File: lib/private/db/adaptersqlite.php (hunk 4 above)
Comment (nickvergessen) [inline lib/private/db/adaptersqlite.php]: Fixed all doc blocks

**Comment by DeepDiver1975** [comment]
:+1: ready to go

**Comment by DeepDiver1975** [comment]
@PVince81 @icewind1991 please review - THX

File: lib/private/db/adapter.php [hunk 5]
```diff
@@ -40,44 +40,37 @@ public function fixupStatement($statement) {
...
+	 * @param string $table The table name (will replace *PREFIX* with the actual prefix)
+	 * @param array $input data that should be inserted into the table  (column name => value)
+	 * @param array|null $compare List of values that should be checked for "if not exists"
+	 *				If this is null or an empty array, all keys of $input will be compared
+	 * @return int number of inserted rows
+	 * @throws \Doctrine\DBAL\DBALException
 	 */
-	public function insertIfNotExist($table, $input) {
+	public function insertIfNotExist($table, $input, array $compare = null) {
+		if ($compare === null) {
```
Comment (nickvergessen) [inline lib/private/db/adapter.php]: this should be `empty()` so we use all keys when an empty array is passed aswell :(, will fix asap

**Comment by DeepDiver1975** [comment]
@nickvergessen shall we squash to allow easier backport?

**Comment by nickvergessen** [comment]
No, I prefer a cherry-pick session over a huge commit that no one understands anymore.

**Comment by nickvergessen** [comment]
Second review @MorrisJobke @LukasReschke ?

File: lib/private/allconfig.php (hunk 3 above)
Comment (MorrisJobke) [inline lib/private/allconfig.php]: Does this really work? The backticks are missing and I remind problems on OCI with this. @DeepDiver1975 Can you run the oracle test suite?

Comment (nickvergessen) [inline lib/private/allconfig.php]: The method adds the backticks. The code around here, does not know anything about sql syntax ;)

Comment (nickvergessen) [inline lib/private/allconfig.php]: The backticks are in `lib/private/db/adapter.php`

**Comment by DeepDiver1975** [comment]
@MorrisJobke @LukasReschke @PVince81 @icewind1991 please review - THX

**Comment by MorrisJobke** [comment]
Works :+1:

**Comment by nickvergessen** [comment]
Backport is at #14914

**Pull request merged**